### PR TITLE
Base BDD test scaffolding

### PR DESCRIPTION
**Pull request by dmitrii-ubskii**
## What is the goal of this PR?

We create the `behaviour` test package and introduce the BDD context structure. We also implement basic connection and database management BDD tests.

## What are the changes implemented in this PR?

We add the typedb-behaviour dependency, and imlpement the database, session, and transaction steps, along with the TypeQL steps necessary to write schema and data.

Drive-by: fix clippy warnings.

### CODE REVIEW

File: dependencies/vaticle/repositories.bzl [hunk 1]
```diff
@@ -24,8 +24,8 @@ load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")
 def vaticle_dependencies():
     git_repository(
         name = "vaticle_dependencies",
-        remote = "https://github.com/vaticle/dependencies",
-        commit = "76636b1672b04e9880439395b8913231724ae459", # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_dependencies
+        remote = "https://github.com/dmitrii-ubskii/dependencies",
+        commit = "cc2090c816b2d3bbdcba8f04bf22b0b718e7b6ce", # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_dependencies
```
Comment (dmitrii-ubskii) [inline dependencies/vaticle/repositories.bzl:28]: Pending https://github.com/vaticle/dependencies/pull/437

File: src/connection/network/transmitter/response_sink.rs [hunk 2]
```diff
@@ -58,7 +58,7 @@ impl<T> ResponseSink<T> {
         }
     }
 
-    pub(super) async fn error(self, error: ConnectionError) {
+    pub(super) fn error(self, error: ConnectionError) {
```
Comment (dmitrii-ubskii) [inline src/connection/network/transmitter/response_sink.rs:61]: Unnecessary `async`.

File: src/connection/runtime.rs [hunk 3]
```diff
@@ -44,7 +44,7 @@ impl BackgroundRuntime {
         thread::Builder::new().name("gRPC worker".to_string()).spawn(move || {
             async_runtime.block_on(async move {
                 shutdown_source.recv().await;
-            })
+            });
```
Comment (dmitrii-ubskii) [inline src/connection/runtime.rs:47]: Not strictly necessary, but more consistent. Adding a semicolon here makes it explicit that the closure returns `()`.

File: src/database/session.rs [hunk 4]
```diff
@@ -124,6 +124,6 @@ impl Session {
             .await?;
 
         *self.server_session_info.write().unwrap() = session_info;
-        Transaction::new(transaction_stream)
+        Ok(Transaction::new(transaction_stream))
```
Comment (dmitrii-ubskii) [inline src/database/session.rs:127]: `Transaction::new` is no longer fallible.

File: src/database/transaction.rs [hunk 5]
```diff
@@ -39,15 +39,23 @@ pub struct Transaction<'a> {
...
-        Ok(Transaction {
+        Transaction {
             type_: transaction_stream.type_(),
             options: transaction_stream.options().clone(),
             query: QueryManager::new(transaction_stream.clone()),
             transaction_stream,
             _lifetime_guard: PhantomData::default(),
-        })
+        }
+    }
```
Comment (dmitrii-ubskii) [inline src/database/transaction.rs:51]: Infallible `new()` now reflected in the signature.

File: tests/behaviour.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+use typedb_client::{Connection, Database, DatabaseManager, Transaction};
+
+use self::steps::session_tracker::SessionTracker;
+
+#[derive(Debug, World)]
+pub struct TypeDBWorld {
+    pub connection: Connection,
+    pub databases: DatabaseManager,
+    pub session_trackers: Vec<SessionTracker>,
+}
```
Comment (dmitrii-ubskii) [inline tests/behaviour.rs:35]: Context struct for BDD tests.

File: tests/behaviour.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    {$($(#[step($pattern:expr)])+ $async:ident fn $fn_name:ident $args:tt $body:tt)+} => {
+        $($(
+        #[given($pattern)]
+        #[when($pattern)]
+        #[then($pattern)]
+        )*
+        $async fn $fn_name $args $body
+        )*
+    };
+}
```
Comment (dmitrii-ubskii) [inline tests/behaviour.rs:84]: This is a band-aid over us not being 100% consistent in the step definitions. This way, a `#[step(expr)]` annotation becomes admissible in the `Given`, `When`, and `Then` steps.

File: tests/steps/connection/session/steps.rs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    async fn sessions_have_databases(world: &mut TypeDBWorld, step: &Step) {
+        assert_eq!(step.table.as_ref().unwrap().rows.len(), world.session_trackers.len());
+        for (name, session_tracker) in util::iter_table(step).zip(&world.session_trackers) {
+            assert_eq!(name, session_tracker.session().database_name());
+        }
+    }
+
+    #[step(expr = "set session option {word} to: {word}")]
+    async fn set_session_option_to(_world: &mut TypeDBWorld, _option: String, _value: String) {
+        todo!()
```
Comment (flyingsilverfin) [inline tests/steps/connection/session/steps.rs:102]: hanging TODO

File: tests/behaviour/mod.rs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            .max_concurrent_scenarios(Some(1))
+            .with_default_cli()
+            .after(|_, _, _, _, world| {
+                Box::pin(async {
+                    try_join_all(world.unwrap().databases.all().await.unwrap().into_iter().map(Database::delete))
+                        .await
+                        .unwrap();
+                })
+            })
+            .filter_run(glob, |_, _, sc| !sc.tags.iter().any(|t| t == "ignore" || t == "ignore-typedb"))
```
Comment (flyingsilverfin) [inline tests/behaviour/mod.rs:54]: you're missing two more:
`ignore-typedb-client-rust` and `ignore-client-rust`

File: tests/behaviour/mod.rs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        !Self::cucumber()
+            .repeat_failed()
+            .fail_on_skipped()
+            .max_concurrent_scenarios(Some(1))
+            .with_default_cli()
+            .after(|_, _, _, _, world| {
+                Box::pin(async {
+                    try_join_all(world.unwrap().databases.all().await.unwrap().into_iter().map(Database::delete))
+                        .await
+                        .unwrap();
```
Comment (flyingsilverfin) [inline tests/behaviour/mod.rs:51]: The "meat" of this can be its own method called `afterScenario()`

File: tests/behaviour/typeql/steps.rs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+use cucumber::{gherkin::Step, given, then, when};
+use futures::TryStreamExt;
+
+use crate::{behaviour::Context, generic_step_impl};
+
+generic_step_impl! {
+    #[step(expr = "typeql define")]
+    async fn typeql_define(world: &mut Context, step: &Step) {
+        // FIXME parse
+        world.transaction().query().define(step.docstring().unwrap()).await.unwrap();
```
Comment (flyingsilverfin) [inline tests/behaviour/typeql/steps.rs:31]: this is implementable right?

Comment (flyingsilverfin) [reply tests/behaviour/typeql/steps.rs:31]: and below

File: tests/behaviour/connection/steps.rs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+ * under the License.
+ */
+
+use cucumber::{given, then, when};
+
+use crate::{behaviour::Context, generic_step_impl};
+
+generic_step_impl! {
+    #[step("connection has been opened")]
+    async fn connection_has_been_opened(_: &mut Context) {}
```
Comment (flyingsilverfin) [inline tests/behaviour/connection/steps.rs:28]: Pretty interesting - this is a Noop because all held connections are open?

**Review by flyingsilverfin — APPROVED** [review]
Please fix minor comments before merging

File: tests/behaviour/typeql/steps.rs (hunk 11 above)
Comment (dmitrii-ubskii) [reply tests/behaviour/typeql/steps.rs]: @dmikhalin is already working on these fixmes in his fork.

**Pull request merged**